src/libgap-api.c

```diff
@@ -0,0 +1,337 @@
+#include "stringobj.h"
+
+Obj FuncREAD_ALL_COMMANDS(
+    Obj self, Obj instream, Obj echo, Obj capture, Obj resultCallback);
+
+// Setup and initialisation
+void GAP_Initialize(int          argc,
+                    char **      argv,
+                    char **      env,
+                    CallbackFunc markBagsCallback,
+                    CallbackFunc errorCallback)
+{
+    InitializeGap(&argc, argv, env);
+    SetExtraMarkFuncBags(markBagsCallback);
+    STATE(JumpToCatchCallback) = errorCallback;
+}
+
+// TODO: This function is far from completely finalizing a GAP session
+//       it potentially has to close all open file descriptors, kill
+//       subprocesses, etc
+void GAP_Finalize(void)
+{
+    FinishBags();
+}
+
+void GAP_ClearError(void)
+{
+    ClearError();
+}
+
+void GAP_Collect(void)
+{
+    CollectBags(0, 1);
+}
+
+const char * GAP_GetTNam(Obj val)
+{
+    return TNAM_OBJ(val);
+}
+
+uint64_t GAP_GetTNum(Obj val)
+{
+    return TNUM_OBJ(val);
+}
+
+
+//
+// Booleans
+//
+bool GAP_IsBoolean(Obj val)
+{
+    return TNUM_OBJ(val) == T_BOOL;
+}
+
+// What about fail :/
+Obj GAP_MakeBoolean(bool val)
+{
+    return val == true ? True : False;
+}
+
+// Should the Get functions assert? get a pointer to put the return value
+// into?
+// If we return 0/1/2 then "Fail" becomes a "true" C value
+int GAP_GetBoolean(Obj val, int * res)
+{
+    GAP_ASSERT(res != NULL);
+    if (val == True)
+        *res = 1;
+    else if (val == False)
+        *res = 0;
+    else if (val == Fail)
+        *res = 2;
+    else
+        *res = -1;
+}
+
+inline Obj GAP_True()
+{
+    return True;
+}
+
+inline Obj GAP_False()
+{
+    return False;
+}
+
+inline Obj GAP_Fail()
+{
+    return Fail;
+}
+
+
+// TODO: depending on 32/64bit? Int?
+bool GAP_IsImmediateInteger(Obj val)
+{
+    return IS_INTOBJ(val);
+}
+
+Obj GAP_MakeImmediateInteger(int64_t val)
+{
+    INTOBJ_INT(val);
+}
+
+int GAP_GetImmediateInteger(Obj val, uint64_t * res)
+{
+    GAP_ASSERT(res);
+    res = INT_INTOBJ(val);
+}
+
+// What about Immediate FFE elements?
+
+//
+// Integers
+//
+bool GAP_IsInteger(Obj val)
+{
+    return IS_INT(val);
+}
+
+Obj GAP_MakeInteger(mpz_t val)    // or something.
+{
+}
+
+
+size_t GAP_GetIntegerSize(Obj val)
+{
+}
+
+// or something, either needs allocation or a way of knowing how much space is
+// needed
+int GAP_GetInteger(Obj val, mpz_t res)
+{
+    GAP_ASSERT(res);
+}
+
+//
+// Strings
+//
+bool GAP_IsString(Obj val)
+{
+    return IS_STRING(val);
+}
+
+Obj GAP_MakeString(const char * str)
+{
+    return MakeString(str);
+}
+
+size_t GAP_LengthString(
+    Obj string)    // Document which length we mean (including/excluding \0)!
+{
+    // assert this is a string?
+    // Or test and error?
+    return GET_LEN_STRING(string);
+}
+
+size_t GAP_LengthCString(
+    Obj string)    // Document which length we mean (including/excluding \0)!
+{
+    // assert this is a string?
+    // Or test and error?
+    return GET_LEN_STRING(string) + 1;
```

Comment:
I find this function extremely confusing. From the name, I'd expect it be "strlen", but it isn't. I can't think of a useful application either.

So I'd just remove it, instead of trying to figure out how to do it right